# 과제 1 DF 추출

In [3]:
import pandas as pd
import numpy as np

In [18]:
url = 'https://ko.wikipedia.org/wiki/%ED%83%9C%ED%92%8D#%ED%83%9C%ED%92%8D%EC%97%90_%EB%8C%80%ED%95%9C_%ED%86%B5%EA%B3%84'
dft = pd.read_html(url)
len(dft)


25

In [24]:
dft[3]

태풍 발생이 많았던 해                   태풍 발생이 적었던 해            
            순위  발생수           연도           순위  발생수     연도
0           1위  39개        1967년           1위  14개  2010년
1           2위  36개  1994년 1971년           2위  16개  1998년
2           4위  35개        1966년           3위  19개  1969년

In [33]:
dfe = pd.read_html('https://ko.wikipedia.org/wiki/%EC%97%90%EB%A6%AD_%ED%81%B4%EB%9E%A9%ED%8A%BC%EC%9D%98_%EC%9D%8C%EB%B0%98_%EB%AA%A9%EB%A1%9D')#, na_values='—')
dfe[2]

발매년도                                              음반 정보 각 국의 차트 순위      \
    발매년도                                              음반 정보         UK AUS   
0   1970           Eric Clapton 레이블: Polydor 발매일: July 1970         14   7   
1   1974  461 Ocean Boulevard Label: RSO Released: July ...          3   2   
2   1975  There's One in Every Crowd 레이블: RSO 발매일: March...         15   —   
3   1976         No Reason to Cry 레이블: RSO 발매일: August 1976          8   —   
4   1977               Slowhand 레이블: RSO 발매일: November 1977         23   —   
5   1978           Backless 레이블: Polydor 발매일: November 1978         18   —   
6   1981     Another Ticket 레이블: Polydor 발매일: February 1981         18   —   
7   1983  Money and Cigarettes 레이블: Duck · Warner 발매일: F...         13   —   
8   1985  Behind the Sun 레이블: Duck · Warner 발매일: March 1985          8   —   
9   1986       August 레이블: Duck · Warner 발매일: November 1986          3   —   
10  1989   Journeyman 레이블: Duck · Warner 발매일: November 1989          2  27   
11  1994  From the Cradle 레이블: Duck · Warner 발매일: Septem...          1   6   
12  1998               Pilgrim 레이블: Reprise 벌매일: March 1998          3  17   
13  2001               Reptile 레이블: Reprise 발매일: March 2001          7  20   
14  2004    Me and Mr. Johnson 레이블: Reprise 발매일: March 2004         10  23   
15  2004  Sessions for Robert J 레이블: Reprise 발매일: Decemb...          –   –   
16  2005            Back Home 레이블: Reprise 발매일: August 2005         19   –   
17  2010           Clapton 레이블: Reprise 발매일: September 2010          7  21   
18  2013              Old Sock 레이블: Reprise 발매일: March 2013         13  22   
19  2016  I Still Do 레이블: Bushbranch · Surfdog 발매일: May ...          6  10   
20  2018  Happy Xmas 레이블: Bushbranch · Surfdog 발매일: 12 O...         97   –   

                                                        
      BEL CAN DEN   FRA     ITA  JPN     NED SWE    US  
0       —   —   —     —       —   85       —   —    13  
1       —   1   1     —      24    8       4   —     1  
2       —   —   —    15       —   31       —   —    21  
3       —   —   —     —       —   11       9   —    15  
4      89   3  19    13       —   35      17  41     2  
5       —   —   —    21       —   53      22  28     8  
6       —   6   7     9       —   32      38   —     7  
7       —   —   —     —       —    3      16   5    16  
8       —  10   —     —       —   14       2  10    34  
9       —  40   —     —      17   12      29   7    37  
10      —   7   —    17      11   29       9   3    16  
11      6   2   2     6      10    6       3   2     1  
12      6   5   6     5       3    1       7   3     4  
13     10  11   7     9       8    6       9   9     5  
14     12   3   9    11       9    8      12   7     6  
15      –   –   –   120       –  189       –   –   172  
16     11   –   4    16       4    4       9   5    13  
17      8   7   3    10       6    8       9   5     6  
18     11  12   9    21      11    9       7  14     7  
19      5  11  31  9[7]      12   10       5  13  6[8]  
20  64[9]  97   –     –  75[10]    –  111[9]   –    84

# 과제2 메모리 줄이기

In [4]:
movies = pd.read_csv('programing3/data/movie.csv')
movies

color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
1     Color     Gore Verbinski                   302.0     169.0   
2     Color         Sam Mendes                   602.0     148.0   
3     Color  Christopher Nolan                   813.0     164.0   
4       NaN        Doug Walker                     NaN       NaN   
...     ...                ...                     ...       ...   
4911  Color        Scott Smith                     1.0      87.0   
4912  Color                NaN                    43.0      43.0   
4913  Color   Benjamin Roberds                    13.0      76.0   
4914  Color        Daniel Hsia                    14.0     100.0   
4915  Color           Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                         0.0                   855.0  Joel David Moore   
1                       563.0                  1000.0     Orlando Bloom   
2                         0.0                   161.0      Rory Kinnear   
3                     22000.0                 23000.0    Christian Bale   
4                       131.0                     NaN        Rob Walker   
...                       ...                     ...               ...   
4911                      2.0                   318.0     Daphne Zuniga   
4912                      NaN                   319.0     Valorie Curry   
4913                      0.0                     0.0     Maxwell Moody   
4914                      0.0                   489.0     Daniel Henney   
4915                     16.0                    16.0  Brian Herzlinger   

      actor_1_facebook_likes        gross                           genres  \
0                     1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi   
1                    40000.0  309404152.0         Action|Adventure|Fantasy   
2                    11000.0  200074175.0        Action|Adventure|Thriller   
3                    27000.0  448130642.0                  Action|Thriller   
4                      131.0          NaN                      Documentary   
...                      ...          ...                              ...   
4911                   637.0          NaN                     Comedy|Drama   
4912                   841.0          NaN     Crime|Drama|Mystery|Thriller   
4913                     0.0          NaN            Drama|Horror|Thriller   
4914                   946.0      10443.0             Comedy|Drama|Romance   
4915                    86.0      85222.0                      Documentary   

      ... num_user_for_reviews language  country  content_rating       budget  \
0     ...               3054.0  English      USA           PG-13  237000000.0   
1     ...               1238.0  English      USA           PG-13  300000000.0   
2     ...                994.0  English       UK           PG-13  245000000.0   
3     ...               2701.0  English      USA           PG-13  250000000.0   
4     ...                  NaN      NaN      NaN             NaN          NaN   
...   ...                  ...      ...      ...             ...          ...   
4911  ...                  6.0  English   Canada             NaN          NaN   
4912  ...                359.0  English      USA           TV-14          NaN   
4913  ...                  3.0  English      USA             NaN       1400.0   
4914  ...                  9.0  English      USA           PG-13          NaN   
4915  ...                 84.0  English      USA              PG       1100.0   

      title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0         2009.0                  936.0        7.9          1.78   
1         2007.0                 5000.0        7.1          2.35   
2         2015.0                  393.0        6.8          2.35   
3         2012.0                23000.0        8.5          2.35   
4            NaN                   12.

In [5]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4916 entries, 0 to 4915
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   color                      4897 non-null   object 
 1   director_name              4814 non-null   object 
 2   num_critic_for_reviews     4867 non-null   float64
 3   duration                   4901 non-null   float64
 4   director_facebook_likes    4814 non-null   float64
 5   actor_3_facebook_likes     4893 non-null   float64
 6   actor_2_name               4903 non-null   object 
 7   actor_1_facebook_likes     4909 non-null   float64
 8   gross                      4054 non-null   float64
 9   genres                     4916 non-null   object 
 10  actor_1_name               4909 non-null   object 
 11  movie_title                4916 non-null   object 
 12  num_voted_users            4916 non-null   int64  
 13  cast_total_facebook_likes  4916 non-null   int64

In [6]:
movies.head()

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4    NaN        Doug Walker                     NaN       NaN   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    131.0                     NaN        Rob Walker   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   
3                 27000.0  448130642.0                  Action|Thriller  ...   
4                   131.0          NaN                      Documentary  ...   

  num_user_for_reviews language  country  content_rating       budget  \
0               3054.0  English      USA           PG-13  237000000.0   
1               1238.0  English      USA           PG-13  300000000.0   
2                994.0  English       UK           PG-13  245000000.0   
3               2701.0  English      USA           PG-13  250000000.0   
4                  NaN      NaN      NaN             NaN          NaN   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4         NaN                   12.0        7.1           NaN   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  
3               164000  
4                    0  

[5 rows x 28 columns]

In [7]:
movies.describe().T

count          mean           std      min  \
num_critic_for_reviews     4867.0  1.379889e+02  1.202394e+02     1.00   
duration                   4901.0  1.070908e+02  2.528602e+01     7.00   
director_facebook_likes    4814.0  6.910145e+02  2.832954e+03     0.00   
actor_3_facebook_likes     4893.0  6.312763e+02  1.625875e+03     0.00   
actor_1_facebook_likes     4909.0  6.494488e+03  1.510699e+04     0.00   
gross                      4054.0  4.764451e+07  6.737255e+07   162.00   
num_voted_users            4916.0  8.264492e+04  1.383222e+05     5.00   
cast_total_facebook_likes  4916.0  9.579816e+03  1.816432e+04     0.00   
facenumber_in_poster       4903.0  1.377320e+00  2.023826e+00     0.00   
num_user_for_reviews       4895.0  2.676688e+02  3.729348e+02     1.00   
budget                     4432.0  3.654749e+07  1.002427e+08   218.00   
title_year                 4810.0  2.002448e+03  1.245398e+01  1916.00   
actor_2_facebook_likes     4903.0  1.621924e+03  4.011300e+03     0.00   
imdb_score                 4916.0  6.437429e+00  1.127802e+00     1.60   
aspect_ratio               4590.0  2.222349e+00  1.402940e+00     1.18   
movie_facebook_likes       4916.0  7.348294e+03  1.920602e+04     0.00   

                                  25%          50%          75%           max  
num_critic_for_reviews          49.00       108.00       191.00  8.130000e+02  
duration                        93.00       103.00       118.00  5.110000e+02  
director_facebook_likes          7.00        48.00       189.75  2.300000e+04  
actor_3_facebook_likes         132.00       366.00       633.00  2.300000e+04  
actor_1_facebook_likes         607.00       982.00     11000.00  6.400000e+05  
gross                      5019656.25  25043962.00  61108412.75  7.605058e+08  
num_voted_users               8361.75     33132.50     93772.75  1.689764e+06  
cast_total_facebook_likes     1394.75      3049.00     13616.75  6.567300e+05  
facenumber_in_poster             0.00         1.00         2.00  4.300000e+01  
num_user_for_reviews            64.00       153.00       320.50  5.060000e+03  
budget                     6000000.00  19850000.00  43000000.00  4.200000e+09  
title_year                    1999.00      2005.00      2011.00  2.016000e+03  
actor_2_facebook_likes         277.00       593.00       912.00  1.370000e+05  
imdb_score                       5.80         6.60         7.20  9.500000e+00  
aspect_ratio                     1.85         2.35         2.35  1.600000e+01  
movie_facebook_likes             0.00       159.00      2000.00  3.490000e+05

mean과 max의 길이의 숫자 중 최대로 높은것은 budget의 max값(4.200000e+09) 이기에 float16을 사용해도 될 것 같습니다.

int형의 경우 num_voted_users의 max값(1.689764e+06)이기에 int32를 사용해도 될 것 같습니다.


In [11]:
# 변수명 뒤에 공백이 있을경우 변환이 안됨. 주의
movies_m = pd.read_csv('programing3/data/movie.csv',
                        dtype={
                            'num_critic_for_reviews' : np.float16,
                            'duration' : np.float16,
                            'director_facebook_likes' : np.float16,
                            'actor_3_facebook_likes' : np.float16,
                            'actor_1_facebook_likes' : np.float16,
                            'gross' : np.float16,
                            'num_voted_users' : np.int32,
                            'cast_total_facebook_likes' : np.int32,
                            'facenumber_in_poster' : np.float16,
                            'num_user_for_reviews' : np.float16,
                            'budget' : np.float16,
                            'title_year' : np.float16,
                            'actor_2_facebook_likes' : np.float16,
                            'imdb_score' : np.float16,
                            'aspect_ratio' : np.float16,
                            'movie_facebook_likes' : np.int32
                        })
movies_m.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4916 entries, 0 to 4915
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   color                      4897 non-null   object 
 1   director_name              4814 non-null   object 
 2   num_critic_for_reviews     4867 non-null   float16
 3   duration                   4901 non-null   float16
 4   director_facebook_likes    4814 non-null   float16
 5   actor_3_facebook_likes     4893 non-null   float16
 6   actor_2_name               4903 non-null   object 
 7   actor_1_facebook_likes     4909 non-null   float16
 8   gross                      4054 non-null   float16
 9   genres                     4916 non-null   object 
 10  actor_1_name               4909 non-null   object 
 11  movie_title                4916 non-null   object 
 12  num_voted_users            4916 non-null   int32  
 13  cast_total_facebook_likes  4916 non-null   int32

In [17]:
round(643.4/1126.4, 2)
# 많이 줄어 든 것을 볼 수 있습니다.

0.57

In [9]:
movies_m.head()

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4    NaN        Doug Walker                     NaN       NaN   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23008.0    Christian Bale   
4                    131.0                     NaN        Rob Walker   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   
3                 27000.0  448130642.0                  Action|Thriller  ...   
4                   131.0          NaN                      Documentary  ...   

  num_user_for_reviews language  country  content_rating budget  title_year  \
0               3054.0  English      USA           PG-13    inf      2009.0   
1               1238.0  English      USA           PG-13    inf      2007.0   
2                994.0  English       UK           PG-13    inf      2015.0   
3               2700.0  English      USA           PG-13    inf      2012.0   
4                  NaN      NaN      NaN             NaN    NaN         NaN   

  actor_2_facebook_likes imdb_score  aspect_ratio movie_facebook_likes  
0                  936.0   7.898438      1.780273                33000  
1                 5000.0   7.101562      2.349609                    0  
2                  393.0   6.800781      2.349609                85000  
3                23000.0   8.500000      2.349609               164000  
4                   12.0   7.101562           NaN                    0  

[5 rows x 28 columns]

In [10]:
movies_m.describe(include='O')

color     director_name    actor_2_name genres    actor_1_name  \
count    4897              4814            4903   4916            4909   
unique      2              2397            3030    914            2095   
top     Color  Steven Spielberg  Morgan Freeman  Drama  Robert De Niro   
freq     4693                26              18    233              48   

       movie_title  actor_3_name   plot_keywords  \
count         4916          4893            4764   
unique        4916          3519            4756   
top         Avatar  Steve Coogan  based on novel   
freq             1             8               4   

                                          movie_imdb_link language country  \
count                                                4916     4904    4911   
unique                                               4916       47      65   
top     http://www.imdb.com/title/tt0499549/?ref_=fn_t...  English     USA   
freq                                                    1     4582    3710   

       content_rating  
count            4616  
unique             18  
top                 R  
freq             2067

moives_m 의 object들의 명목형 변수를 보게 되면 color 에 unique 2개, content_rating에 unique 18개가 뜨는 것을 볼 수 있습니다. 다른 항목들은 47, 914, 4916가지로 나뉘기에 나머지는 카테고리를 쓰기에 적합하지 않습니다.

In [20]:
movies_m2 = pd.read_csv('programing3/data/movie.csv',
                        dtype={
                            'num_critic_for_reviews' : np.float16,
                            'duration' : np.float16,
                            'director_facebook_likes' : np.float16,
                            'actor_3_facebook_likes' : np.float16,
                            'actor_1_facebook_likes' : np.float16,
                            'gross' : np.float16,
                            'num_voted_users' : np.int32,
                            'cast_total_facebook_likes' : np.int32,
                            'facenumber_in_poster' : np.float16,
                            'num_user_for_reviews' : np.float16,
                            'budget' : np.float16,
                            'title_year' : np.float16,
                            'actor_2_facebook_likes' : np.float16,
                            'imdb_score' : np.float16,
                            'aspect_ratio' : np.float16,
                            'movie_facebook_likes' : np.int32,
                            'color' : 'category',
                            'content_rating' : 'category'
                        })
movies_m2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4916 entries, 0 to 4915
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype   
---  ------                     --------------  -----   
 0   color                      4897 non-null   category
 1   director_name              4814 non-null   object  
 2   num_critic_for_reviews     4867 non-null   float16 
 3   duration                   4901 non-null   float16 
 4   director_facebook_likes    4814 non-null   float16 
 5   actor_3_facebook_likes     4893 non-null   float16 
 6   actor_2_name               4903 non-null   object  
 7   actor_1_facebook_likes     4909 non-null   float16 
 8   gross                      4054 non-null   float16 
 9   genres                     4916 non-null   object  
 10  actor_1_name               4909 non-null   object  
 11  movie_title                4916 non-null   object  
 12  num_voted_users            4916 non-null   int32   
 13  cast_total_facebook_likes  4916 n

In [26]:
round(577/1126.4, 2)

0.51

# 최종적으로 51% 정도 줄어 든것을 볼 수 있습니다!

In [22]:
movies_m2.head()

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4    NaN        Doug Walker                     NaN       NaN   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23008.0    Christian Bale   
4                    131.0                     NaN        Rob Walker   

   actor_1_facebook_likes  gross                           genres  ...  \
0                  1000.0    inf  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0    inf         Action|Adventure|Fantasy  ...   
2                 11000.0    inf        Action|Adventure|Thriller  ...   
3                 27008.0    inf                  Action|Thriller  ...   
4                   131.0    NaN                      Documentary  ...   

  num_user_for_reviews language  country  content_rating budget  title_year  \
0               3054.0  English      USA           PG-13    inf      2009.0   
1               1238.0  English      USA           PG-13    inf      2007.0   
2                994.0  English       UK           PG-13    inf      2015.0   
3               2700.0  English      USA           PG-13    inf      2012.0   
4                  NaN      NaN      NaN             NaN    NaN         NaN   

  actor_2_facebook_likes imdb_score  aspect_ratio movie_facebook_likes  
0                  936.0   7.898438      1.780273                33000  
1                 5000.0   7.101562      2.349609                    0  
2                  393.0   6.800781      2.349609                85000  
3                23008.0   8.500000      2.349609               164000  
4                   12.0   7.101562           NaN                    0  

[5 rows x 28 columns]

# chunksize를 이용해 데이터 불러오기(최후 방법!)

In [27]:
movies_iter = pd.read_csv('programing3/data/movie.csv',
                         dtype={
                            'num_critic_for_reviews' : np.float16,
                            'duration' : np.float16,
                            'director_facebook_likes' : np.float16,
                            'actor_3_facebook_likes' : np.float16,
                            'actor_1_facebook_likes' : np.float16,
                            'gross' : np.float16,
                            'num_voted_users' : np.int32,
                            'cast_total_facebook_likes' : np.int32,
                            'facenumber_in_poster' : np.float16,
                            'num_user_for_reviews' : np.float16,
                            'budget' : np.float16,
                            'title_year' : np.float16,
                            'actor_2_facebook_likes' : np.float16,
                            'imdb_score' : np.float16,
                            'aspect_ratio' : np.float16,
                            'movie_facebook_likes' : np.int32,
                            'color' : 'category',
                            'content_rating' : 'category'
                        },
                       chunksize=1000
                       )
def process(df):
    return f'processed {df.shape} shape of df {df.iloc[0, :]}'

for chunk in movies_iter:
    print(process(chunk))

processed (1000, 28) shape of df color                                                                    Color
director_name                                                    James Cameron
num_critic_for_reviews                                                   723.0
duration                                                                 178.0
director_facebook_likes                                                    0.0
actor_3_facebook_likes                                                   855.0
actor_2_name                                                  Joel David Moore
actor_1_facebook_likes                                                  1000.0
gross                                                                      inf
genres                                         Action|Adventure|Fantasy|Sci-Fi
actor_1_name                                                       CCH Pounder
movie_title                                                             Avatar
num_voted_users    